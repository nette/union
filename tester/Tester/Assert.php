<?php

/**
 * Nette Framework
 *
 * @copyright  Copyright (c) 2004, 2010 David Grudl
 * @license    http://nette.org/license  Nette license
 * @link       http://nette.org
 * @category   Nette
 * @package    Nette\Test
 */



/**
 * Asseratation test helpers.
 *
 * @author     David Grudl
 * @package    Nette\Test
 */
class Assert
{

	/**
	 * Checks assertation.
	 * @param  mixed  expected
	 * @param  mixed  actual
	 * @return void
	 */
	public static function same($expected, $actual)
	{
		if ($actual !== $expected) {
			self::note('Failed asserting that ' . self::dump($actual) . ' is not identical to ' . self::dump($expected));
		}
	}



	/**
	 * Checks TRUE assertation.
	 * @param  mixed  actual
	 * @return void
	 */
	public static function true($actual)
	{
		if ($actual !== TRUE) {
			self::note('Failed asserting that ' . self::dump($actual) . ' is not TRUE');
		}
	}



	/**
	 * Checks FALSE assertation.
	 * @param  mixed  actual
	 * @return void
	 */
	public static function false($actual)
	{
		if ($actual !== FALSE) {
			self::note('Failed asserting that ' . self::dump($actual) . ' is not FALSE');
		}
	}



	/**
	 * Checks NULL assertation.
	 * @param  mixed  actual
	 * @return void
	 */
	public static function null($actual)
	{
		if ($actual !== NULL) {
			self::note('Failed asserting that ' . self::dump($actual) . ' is not NULL');
		}
	}



	/**
	 * Dumps information about a variable in readable format.
	 * @param  mixed  variable to dump
	 * @return void
	 */
	private static function dump($var)
	{
		if (is_bool($var)) {
			return $var ? 'TRUE' : 'FALSE';

		} elseif ($var === NULL) {
			return "NULL";

		} elseif (is_int($var)) {
			return "$var";

		} elseif (is_float($var)) {
			return "$var";

		} elseif (is_string($var)) {
			return var_export($var, TRUE);

		} elseif (is_array($var)) {
			return "array(" . count($var) . ")";

		} elseif ($var instanceof Exception) {
			return 'Exception ' . get_class($var) . ': ' . ($var->getCode() ? '#' . $var->getCode() . ' ' : '') . $var->getMessage();

		} elseif (is_object($var)) {
			$arr = (array) $var;
			return "object(" . get_class($var) . ") (" . count($arr) . ")";

		} elseif (is_resource($var)) {
			return "resource(" . get_resource_type($var) . ")";

		} else {
			return "unknown type";
		}
	}



	/**
	 * Returns message and file and line from call stack.
	 * @param  string
	 * @return void
	 */
	private static function note($message)
	{
		echo $message;
		$trace = debug_backtrace();
		if (isset($trace[1]['file'], $trace[1]['line'])) {
			echo ' in file ' . $trace[1]['file'] . ' on line ' . $trace[1]['line'];
		}
		echo "\n\n";
	}

}
