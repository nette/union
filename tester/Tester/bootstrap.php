<?php

/**
 * Test environment initialization.
 *
 * @author     David Grudl
 */

require __DIR__ . '/Framework/Helpers.php';
require __DIR__ . '/Framework/Assert.php';
require __DIR__ . '/Framework/TestCase.php';
require __DIR__ . '/Framework/DomQuery.php';
require __DIR__ . '/CodeCoverage/Collector.php';
require __DIR__ . '/Runner/Job.php';


// configure PHP environment
error_reporting(E_ALL | E_STRICT);
ini_set('display_errors', TRUE);
ini_set('html_errors', FALSE);
ini_set('log_errors', FALSE);


// catch errors/warnings/notices
set_error_handler(function($severity, $message, $file, $line) {
	if (($severity & error_reporting()) === $severity) {
		$e = new ErrorException($message, 0, $severity, $file, $line);
		echo "\nError: $message in $file:$line\nStack trace:\n" . $e->getTraceAsString();
		exit(Tester\Runner\Job::CODE_ERROR);
	}
	return FALSE;
});


// catch exceptions
set_exception_handler(function($e) {
	echo "\n" . ($e instanceof Tester\AssertException ? '' : get_class($e) . ': ') . $e->getMessage();
	$trace = $e->getTrace();
	while (isset($trace[0]['file']) && substr($trace[0]['file'], strlen(__DIR__))  === __DIR__) {
		array_shift($trace);
	}
	while ($trace) {
		if (isset($trace[0]['file'], $trace[0]['line'])) {
			echo "\nin " . implode(DIRECTORY_SEPARATOR, array_slice(explode(DIRECTORY_SEPARATOR, $trace[0]['file']), -3)) . ':' . $trace[0]['line'];
		}
		array_shift($trace);
	}
	exit($e instanceof Tester\AssertException ? Tester\Runner\Job::CODE_FAIL : Tester\Runner\Job::CODE_ERROR);
});
