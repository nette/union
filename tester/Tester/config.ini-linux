[php-5.2.0]
binary = "/opt/php-5.2.0/bin/php5-cgi"
php.ini = "/opt/php-5.2.0/etc/php.ini"
libraries = "/opt/php-5.2.0/lib"

[php-5.3.0]
binary = "/opt/php-5.3.0/bin/php5-cgi"
php.ini = "/opt/php-5.3.0/etc/php.ini"
libraries = "/opt/php-5.3.0/lib"
