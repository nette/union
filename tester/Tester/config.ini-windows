[php-5.2.0]
binary = "D:\Www\PHP-5.2.0\php-cgi.exe"
php.ini = "D:\Www\PHP-5.2.0\php.ini"

[php-5.3.0]
binary = "D:\Www\PHP-5.3.0\php-cgi.exe"
php.ini = "D:\Www\PHP-5.3.0\php.ini"

