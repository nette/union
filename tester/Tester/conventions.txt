Test case file name
===================

Nette\<class>.<method()>.<area>.<number>.phpt

Nette\Debug.phpt                - tests for a class's basic behaviour
Nette\Debug.fireLog().phpt      - tests for a method's basic behaviour
Nette\Debug.fireLog().inc       - common code for more test cases
Nette\Debug.fireLog().expect    - expected raw output
Nette\Debug.fireLog().area.phpt - tests for a specified area of class/method

- areas: basic, error, bug#123
- numbers have three digits




Test case file layout
=====================

Sections:

PHP (All test cases are valid PHP files)
	- phpDoc with test case name (optional)
	- require 'NetteTest/initialize.php' (required)
    - dump variables using function dump()
    - __halt_compiler() (required if file contains more sections)


---EXPECT---
	The expected output from the test script. It uses a number
	of substitution tags for strings, spaces, digits, etc:

		%a%    one or more of anything except the end of line characters
		%a?%   zero or more of anything except the end of line characters
		%A%    one or more of anything including the end of line characters
		%A?%   zero or more of anything including the end of line characters
		%s%    one or more white space characters except the end of line characters
		%s?%   zero or more white space characters except the end of line characters
		%S%    one or more of characters except the white space
		%S?%   zero or more of characters except the white space
		%c%    a single character of any sort (except the end of line)
		%d%    one or more digits
		%d?%   zero or more digits
		%i%    signed integer value
		%f%    floating point number
		%h%    one or more HEX digits
		%[..]% reg-exp


---EXPECTHEADERS---
	The expected HTTP headers. Any header specified here must exist
	in the response and have the same value or the test fails.
	Additional headers found in the actual tests while running are ignored.

		Content-type: text/html; charset=UTF-8
		Status: 403 Access Denied