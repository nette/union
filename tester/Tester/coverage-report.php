<?php

/**
 * Code coverage HTML report generator.
 *
 * This file is part of the Nette Tester.
 */


require_once __DIR__ . '/CodeCoverage/ReportGenerator.php';


echo '
Code coverage HTML report generator
-----------------------------------
';

$options = (array) getopt('c:s:o:h', array('help'));

if (!$options) { ?>
Usage:
	php coverage-report.php [options]

Options:
	-c <path>  coverage.dat file (default is coverage.dat)
	-s <path>  directory with source code
	-o <path>  output file (default is coverage.html)

<?php
}



$options += array(
	'c' => 'coverage.dat',
	's' => NULL,
	'o' => 'coverage.html',
);

try {
	$generator = new ReportGenerator($options['c'], $options['s']);
	if ($options['o'] === '-') {
		$generator->render();
	} else {
		echo "Generating report to $options[o]\n";
		$generator->render($options['o']);
		echo "Done.\n";
	}

} catch (Exception $e) {
	echo "Fatal error: {$e->getMessage()}\n";
	die(254);
}
