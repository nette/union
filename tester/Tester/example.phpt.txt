Test case file *.phpt
=====================

<?php

/**
 * Test: some test name
 *
 * @author     John Doe
 * @phpVersion < 5.3   default operator is >=
 * @skip       some reason why test is skipped
 * @phpIni     short_open_tag=on
 * @keepTrailingSpaces
 */

require __DIR__ . '/NetteTest/TestHelpers.php';
TestHelpers::startup();

...test code...

	test variables using Assert::*()


	Assert::match() uses a number of substitution tags for strings, spaces, digits, etc:

		%a%    one or more of anything except the end of line characters
		%a?%   zero or more of anything except the end of line characters
		%A%    one or more of anything including the end of line characters
		%A?%   zero or more of anything including the end of line characters
		%s%    one or more white space characters except the end of line characters
		%s?%   zero or more white space characters except the end of line characters
		%S%    one or more of characters except the white space
		%S?%   zero or more of characters except the white space
		%c%    a single character of any sort (except the end of line)
		%d%    one or more digits
		%d?%   zero or more digits
		%i%    signed integer value
		%f%    floating point number
		%h%    one or more HEX digits
		%ns%   PHP namespace
		%[..]% reg-exp
