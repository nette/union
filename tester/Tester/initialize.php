<?php

/**
 * Test initialization and helpers.
 *
 * @author     David Grudl
 * @package    Nette\Test
 */

require dirname(__FILE__) . '/NetteTestCase.php';

require dirname(__FILE__) . '/NetteTestHelpers.php';

require dirname(__FILE__) . '/../../Nette/loader.php';



NetteTestHelpers::startup();



/**
 * Dumps information about a variable in readable format.
 * @param  mixed  variable to dump
 * @param  string
 * @return mixed  variable itself or dump
 */
function dump($var, $message = NULL)
{
	if ($message) {
		echo $message . (preg_match('#[.:?]$#', $message) ? ' ' : ': ');
	}

	NetteTestHelpers::dump($var, 0);
	echo "\n";
	return $var;
}



/**
 * Writes new message.
 * @param  string
 * @return void
 */
function output($message = NULL)
{
	echo $message ? "$message\n\n" : "===\n\n";
}
