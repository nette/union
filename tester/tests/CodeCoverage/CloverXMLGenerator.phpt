<?php

use Tester\Assert;
use Tester\CodeCoverage;

require __DIR__ . '/../bootstrap.php';
require __DIR__ . '/../../src/CodeCoverage/PhpParser.php';
require __DIR__ . '/../../src/CodeCoverage/Generators/AbstractGenerator.php';
require __DIR__ . '/../../src/CodeCoverage/Generators/CloverXMLGenerator.php';


$coveredFile = __DIR__ . DIRECTORY_SEPARATOR . 'CloverXMLGenerator.covered.php';

$coverageData = Tester\FileMock::create(serialize(array(
	$coveredFile => array_map('intval', preg_filter(
		'~.*# (-?\d+)~',
		'$1',
		explode("\n", "\n" . file_get_contents($coveredFile))
	)),
)));

$generator = new CodeCoverage\Generators\CloverXMLGenerator($coverageData, $coveredFile);
$generator->render($output = Tester\FileMock::create('', 'xml'));

Assert::matchFile(__DIR__ . '/CloverXMLGenerator.expected.xml', file_get_contents($output));
