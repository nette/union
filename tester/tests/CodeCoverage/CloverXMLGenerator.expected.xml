<?xml version="1.0"?>
<coverage generated="%d%">
  <project timestamp="%d%">
    <metrics packages="0" files="1" loc="38" ncloc="22" classes="1" methods="3" coveredmethods="2" statements="5" coveredstatements="3" elements="8" coveredelements="5" conditionals="0" coveredconditionals="0"/>
    <file name="%a%CloverXMLGenerator.covered.php">
      <metrics loc="38" ncloc="22" classes="1" methods="3" coveredmethods="2" statements="5" coveredstatements="3" elements="8" coveredelements="5" conditionals="0" coveredconditionals="0"/>
      <class namespace="CloverXMLGeneratorTest" name="Logger">
        <metrics methods="3" coveredmethods="2" statements="5" coveredstatements="3" conditionals="0" coveredconditionals="0" elements="8" coveredelements="5"/>
      </class>
      <line num="19" type="stmt" count="1"/>
      <line num="20" type="stmt" count="1"/>
      <line num="26" type="stmt" count="1"/>
      <line num="32" type="stmt" count="0"/>
      <line num="33" type="stmt" count="0"/>
      <line num="37" type="stmt" count="1"/>
      <line num="38" type="stmt" count="1"/>
    </file>
  </project>
</coverage>
